File: go/Word_Pattern.go
package main

import (
    "fmt"
    "strings"
)

func wordPattern(pattern string, str string) bool {
    arr := strings.Split(str, " ")
    lib := make(map[string]string)

    if len(arr) != len(pattern) {
        return false
    }
    for i, l := 0, len(arr); i < l; i ++ {
        key := string(pattern[i])
        item := arr[i]
        if value, ok := lib["key_" + key]; ok && value != item {
            return false
        }

        if value, ok := lib["item_" + item]; ok && value != key {
            return false
        }
        lib["key_" + key] = item
        lib["item_" + item] = key
    }
    return true
}

func main() {
    pattern := "aaa"
    str := "aa aa aa aa"
    fmt.Println(wordPattern(pattern, str))
}

File: js/Word_Pattern.js
/**
 * @param {string} pattern
 * @param {string} str
 * @return {boolean}
 */
var wordPattern = function(pattern, str) {
    let arr = str.split(' ');
    // 长度不相等
    if (pattern.length !== arr.length) {
        return false;
    }
    let map = {};
    let i = 0;
    let l = arr.length;
    for (; i < l; i ++) {
        let key = pattern[i];
        let item = arr[i];
        if (map['key_' + key] && map['key_' + key] !== item) {
            return false;
        }
        if (map['item_' + item] && map['item_' + item] !== key) {
            return false;
        }
        map['key_' + key] = item;
        map['item_' + item] = key;
    }

    return true;
};

var pattern = 'abba';
var str = 'cat dog dog cat';
console.log(wordPattern(pattern, str));
